### PR TITLE
implement sizeof and alignof operator

### PR DESCRIPTION
**Pull request by krux02**
Don't pull yet. It's almost done. I only want to see if I already broke something by running the tests on the server.

This is the issue I am working on https://github.com/nim-lang/Nim/issues/5493

### CODE REVIEW

**Comment by zah** [comment]
I've always had a paranoid idea for implementing this feature btw.

You can also modify the codegen to introduce a static assert that Nim's assumption was right. Below each C type declaration produced by Nim, you can use the `NIM_CHECK_SIZE` macro from nimbase.h to verify your calculation is working correctly (if we don't run into issue in the next few months, we'll disable these checks).

**Comment by krux02** [comment]
Well I checked how the alignment of C structs actually behave. primitive types get self aligned, and structs get aligned to it's biggest alignment of a member. Compilers may not behave different here, because otherwise headers would not work for interfaces when they are compiled to a different memory layout.

But there are ways to change the behaviour of alignment. I think that when I am already doing this, there should be an align pragma that allows me to set alignment of any type manually.

**Comment by arnetheduck** [comment]
keep in mind that alignment changes with os, architecture and compiler options - here's a telling example:

> A double (eight bytes) will be 8-byte aligned on Windows and 4-byte aligned on Linux (8-byte with -malign-double compile time option).

https://en.wikipedia.org/wiki/Data_structure_alignment#Architectures

**Comment by arnetheduck** [comment]
there's also a feature `-d:checkAbi` if you want to compare what nim thinks with what C thinks for `importc` types - unfortunately the nim standard library is severly lacking in this regard and many crucial structs are simply wrong from an abi-point-of-view

https://github.com/rust-lang/rust/tree/master/src/librustc_back/target is a good reference for common layouts - look for the data_layout string and the docs in http://llvm.org/docs/LangRef.html#data-layout

**Comment by krux02** [comment]
@arnetheduck Thanks for the info, I will take that into consideration. Currently I am on Linux (64-bit) and here the double is 8 byte aligned, so I would not have noticed. I think the best thing to do here would be an alignment map (just a few constants that can be initialized with a few ``when`` statements.

File: compiler/types.nim [hunk 1]
```diff
@@ -1185,173 +1185,377 @@ const
...
-    n.sym.offset = int(currOffset)
+    n.sym.typ.computeSizeAlign
+    result = n.sym.typ.align
+
   else:
-    a = 1
-    result = szNonConcreteType
+    result = 1
+
+proc computeObjectOffsetsRecursive(n: PNode, initialOffset: BiggestInt): tuple[size, align: BiggestInt] =
```
Comment (Araq) [inline compiler/types.nim]: This is now enough code that it should reside it its own module. Suggested name: ``sizeofimpl.nim``

Comment (krux02) [inline compiler/types.nim]: where should that file be? same folder as types.nim, and then just include it from types.nim?

**Comment by krux02** [comment]
I got a very weird problem:

In [computeObjectOffsetsRecursive](https://github.com/krux02/Nim/blob/sizeof-alignof/compiler/types.nim#L1267) I print the result, to check if it has the correct value, and in fact it has. But when I print the exact same values from 
[where I called the function](https://github.com/krux02/Nim/blob/sizeof-alignof/compiler/types.nim#L1483) 

This is the output I get:
```
result.size: 16 result.align: 8
       size: 24        align: 8
```

I get different values. How is that possible? I made sure that the output is only printed for one single type, where I know that the result is wrong.

this is the current testfile in the root of the project: `./koch temp c -r testsizeof`

**Comment by krux02** [comment]
ok, I got the problem fixed. There was one layer of recursion more of the same function that I thought that there would be and that changed the result. Now a lot more tests actually pass.

File: compiler/types.nim (hunk 1 above)
Comment (Araq) [inline compiler/types.nim]: module would be nicer than include file and it should have many dependencies (ast.nim, and not much else?)

Comment (krux02) [inline compiler/types.nim]: well I guess that would be possible

**Comment by Araq** [comment]
What's the status of this? Ready to be reviewed? (I might need this feature soon.)

**Comment by krux02** [comment]
I thought so, but then I realized I am missing out on the ``packed`` pragma. That is the latest TODO I added.   But there is one question that I have. I realized that enums have always an alignment of 2 bytes, no matter the size. Is that intentional? I thought an enum of the size of an int should also have the alignment of an int.

**Comment by Araq** [comment]
>  I realized that enums have always an alignment of 2 bytes, no matter the size. Is that intentional? I thought an enum of the size of an int should also have the alignment of an int.

That's a bug. Most enums fit in a single byte and so should get "byte alignment" (i.e. no alignment). An int sized enum should be aligned like an int, yes.

**Comment by jangko** [comment]
recently I stumbled upon something like this in windows related header files of mingw-gcc.
```C
#include <pshpack2.h>
  typedef struct tagBITMAPFILEHEADER {
    WORD bfType;
    DWORD bfSize;
    WORD bfReserved1;
    WORD bfReserved2;
    DWORD bfOffBits;
  } BITMAPFILEHEADER,*LPBITMAPFILEHEADER,*PBITMAPFILEHEADER;
#include <poppack.h>
```

gcc documentation says those kind of struct alignment exist to provide compatibility for Microsoft's compiler. gcc itself do not use them in it's own header files, only in windows related header files.

pshpackXX force the compiler to use certain alignment of 1,2,4, and 8.

As far as I know, Nim compiler only support byte alignment with `packed` pragma (I might be wrong).

clang compiler usually have compatible features with gcc although it is not in C standard.

icc also support this feature because it claims source and binary compatibility with msvc compiler.

I think if Nim want to support this feature, this PR is the right place to do it since sizeof, alignof and offsetof all deals with alignment.

alignment up to 32 bytes might be needed with AVX instructions.

@araq, @krux02, what do you think?

**Comment by krux02** [comment]
I don't know of the ``pshpack2`` header and how it works. But since it's a non-standard C feature, I think there is no need to support it. That means that wrappers for that non-standard struct should not expose the struct as an object but make the fields accessable via getters/setters. Or use the packed pragma und add alignment bytes manually. An align pragma I think should be supported. I will see what I can do.

**Comment by jangko** [comment]
pshpackXX and poppack contains something like this:

#pragma pack(push,2)

#pragma pack(pop)

File: compiler/semmagic.nim [hunk 2]
```diff
@@ -186,15 +186,60 @@ proc isStrangeArray(t: PType): bool =
...
+    var dotExpr: PNode
+
+    block findDotExpr:
+      if n[1].kind == nkDotExpr:
+        dotExpr = n[1]
+      elif n[1].kind == nkCheckedFieldExpr:
+        dotExpr = n[1][0]
+      else:
+        #echo "evaluating: ", n, " type1: ", type1, " type2: ", type2
+        debug n[1]
```
Comment (Araq) [inline compiler/semmagic.nim]: debug code left over.

**Comment by Parashurama** [comment]
I checked-out the PR locally, here are some results running tests/misc/tsizeof.nim: (Linux x64bits)

I could get any useful results with **``packed``** pragma (some debug ouput).

**RESULTS ***without*** {.packed.}**
running ``nim -r c --cc:gcc tests/misc/tsizeof.nim:``
```
TrivialType:           size:    3 !=     4 align:    1 at line tsizeof.nim(233,18)
InheritanceB.b offset: 16 != 10 at line tsizeof.nim(263,15)
InheritanceC.c offset: 24 != 12 at line tsizeof.nim(264,15)
```
running ``nim -r c --cc:gcc --cpu:i386 --passc: -m32 --passL: -m32  tests/misc/tsizeof.nim:``
```
TrivialType:           size:    3 !=     4 align:    1 at line tsizeof.nim(233,18)
SimpleAlignment:       size:   12 !=    16 align:    8 at line tsizeof.nim(233,18)
AlignAtEnd:            size:   12 !=    16 align:    8 at line tsizeof.nim(233,18)
SimpleBranch:          size:   12 !=    16 align:    8 at line tsizeof.nim(233,18)
PaddingBeforeBranchA:  size:   12 !=    16 align:    8 at line tsizeof.nim(233,18)
RecursiveStuff:        size:   28 !=    40 align:    8 at line tsizeof.nim(233,18)
SimpleAlignment.c offset: 4 != 8 at line tsizeof.nim(241,15)
SimpleBranch.a offset: 4 != 8 at line tsizeof.nim(246,15)
SimpleBranch.b offset: 4 != 8 at line tsizeof.nim(247,15)
SimpleBranch.c offset: 4 != 8 at line tsizeof.nim(248,15)
PaddingBeforeBranchA.a offset: 4 != 8 at line tsizeof.nim(251,15)
Bazing.a offset: 4 != 8 at line tsizeof.nim(260,15)
InheritanceB.b offset: 8 != 6 at line tsizeof.nim(263,15)
InheritanceC.c offset: 12 != 8 at line tsizeof.nim(264,15)
RecursiveStuff.a offset: 4 != 8 at line tsizeof.nim(276,15)
RecursiveStuff.b offset: 4 != 8 at line tsizeof.nim(277,15)
RecursiveStuff.kind2 offset: 4 != 8 at line tsizeof.nim(278,15)
RecursiveStuff.ca1 offset: 8 != 16 at line tsizeof.nim(279,15)
RecursiveStuff.ca2 offset: 12 != 20 at line tsizeof.nim(280,15)
RecursiveStuff.cb offset: 8 != 16 at line tsizeof.nim(281,15)
RecursiveStuff.cc offset: 8 != 16 at line tsizeof.nim(282,15)
RecursiveStuff.d1 offset: 16 != 24 at line tsizeof.nim(283,15)
RecursiveStuff.d2 offset: 20 != 32 at line tsizeof.nim(284,15)
```
running ``nim -r c --cc:clang tests/misc/tsizeof.nim:``
```
TrivialType:           size:    3 !=     4 align:    1 at line tsizeof.nim(233,18)
InheritanceB.b offset: 16 != 10 at line tsizeof.nim(263,15)
InheritanceC.c offset: 24 != 12 at line tsizeof.nim(264,15)
```

**Comment by krux02** [comment]
@Parashurama thanks for the help. I am pretty sure where the problem is. Depending on the architecture and the platform the alignment might not be self aligned for all primitive types. My implementation does not handle those cases properly. Here is where the branching needs to take place:

https://github.com/nim-lang/Nim/pull/5664/files#diff-76f3d75559ee07c14682d1bd5f58e5f9R202

Here some branching based on the platform needs to happen, to get the correct alignment for the different types.

**Comment by krux02** [comment]
Well this pull request is generally correct. The only problem it is facing now is that the alignment values for the builtin types are not set correctly by the compiler as described here:
https://github.com/nim-lang/Nim/issues/6094

**Comment by Parashurama** [comment]
Well, alignment issues appear to be fixed. Can you update your PR?

**Comment by krux02** [comment]
aren't the commits all squashed when merged by default? So where is the advantage when I squash them?

**Comment by dom96** [comment]
They're not. Not by default at least.

**Comment by krux02** [comment]
Well I don't see this as my fault:
```
fatal: unable to access 'https://github.com/nim-lang/csources/': Failed to connect to github.com port 443: Timed out
```

**Comment by Parashurama** [comment]
You should trigger a Travis rebuild with a no-opt amended commit and git push --force-with-lease

**Comment by krux02** [comment]
Well, the test are failing quite regularly because of things that are not my fault. Often it is a timeout somewhere.

**Comment by krux02** [comment]
seems like I still have something to do:

```nim
Failure: reOutputsDiffer
Expected:


Gotten:
tsizeof.nim(240, 10)  alignof SimpleAlignment 8 != 16
TrivialType:           size:    3 !=     4 align:    1
InheritanceB.b offset: 16 != 10
InheritanceC.c offset: 24 != 12
```

**Comment by dom96** [comment]
@krux02 Do consider running the tests locally if travis isn't reliable.

**Comment by Araq** [comment]
This breaks ``async`` on OSX at least --- the problem is that importc'ed structs can be incomplete, covering only the fields that we need. The logic of "cannot evaluate sizeof at compiletime" needs to be preserved and enabled for importc'ed structs.

**Comment by krux02** [comment]
@Araq How should I encode the "cannot evaluate sizeof at compiletime"? And how is it encoded currently?

**Comment by Araq** [comment]
> How should I encode the "cannot evaluate sizeof at compiletime"? And how is it encoded currently?

Check the devel code, size == -1 iirc.

**Comment by krux02** [comment]
@Araq Well ``size = -1`` is already the code for "size not yet computed". And iirc ``size = -2`` is illegal recursion. I will look further into this, after this weekend.

**Comment by yglukhov** [comment]
Hey guys, any news on this?

**Comment by krux02** [comment]
It is definitively not dead. There is still wish for me to implement it properly. But it surely is much harder than I thought it would be and I haven't done anything on it for quite a while. Just the switch between windows and linux arleady has an effect in the alignment. The 64bit integer types is on some platforms 32bit aligned on others 64bit. If you want to crush your head for a while I can rebase this with the latest development branch and then you can inspect the code if you want.

**Comment by arnetheduck** [comment]
maybe some pars of this can be merged already while the tricky parts are being worked on?

**Comment by mratsim** [comment]
Sparked again by @timotheecour, I really want this in v1.

I have `sizeof` workarounds in many projects.

**Comment by krux02** [comment]
@mratsim This gcc compiler flag kills all the problems but with the providid disadvantage. I think we should go that route.


    -malign-double
    -mno-align-double
        Control whether GCC aligns "double", "long double", and "long long" variables on a two-word boundary or a one-word boundary.  Aligning "double" variables on a two-word
        boundary produces code that runs somewhat faster on a Pentium at the expense of more memory.

        On x86-64, -malign-double is enabled by default.

        Warning: if you use the -malign-double switch, structures containing the above types are aligned differently than the published application binary interface specifications
        for the x86-32 and are not binary compatible with structures in code compiled without that switch.

**Comment by mratsim** [comment]
As far as I know, sizeof has been working for a while but alignment doesn't right?

If so, can't we split the PR in `sizeof` that we can commit and `alignof`.

I'm afraid `mno-align-double` will have unforeseen consequences on C wrappers which works on collections or arrays of float64.

**Comment by krux02** [comment]
@mratsim ``sizeof`` depends on ``alignof``, so I am afraid that doesn't work. If anything I would use ``-malign-double``, and no I don't think it would have any consequences on array and seq of float. It would only affect the amount of padding put into this tuple ``(byte,float64)``

**Comment by arnetheduck** [comment]
there are several use cases for sizeof and alignof - one is to casually allocate memory or local within-nim optimizations and customization, but another important use case is interop with other languages and environments.

for the first use case, it makes sense to keep freedom in the compiler to lay things out as it wants to - for the latter it must conform to a particular ABI.

Here's a few examples of 
Thus, I'd say to be feature-complete, the sizeof/alignof features of nim must come with a capability to specify which abi is being targeted - specifically, if, for a particular type, it needs to conform to (most commonly) the `C` abi of the local platform, if it should used a `packed` ABI for network, or whether the compiler should be free to optimize as it sees fit (ie reorder object/tuple fields for optimal packing etc)..
the last one should arguably be the default, but it should also be possible to constrain it further by adding `{.abi: C.}` or {.abi: packed.}`.

Currently, from what I can tell, the generated layout is mostly c-compatible by happy accident, except it breaks when using certain features (at least I can't find anything in the manual) - the manual doesn't say anything, except for the `{.packed.}` pragma which vaguely hints at the problem but remains incomplete and requires magic knowledge of the compiler internals (ie what happens during inheritance, variants etc) to be used safely.

**Comment by krux02** [comment]
``{.packed.}`` simply means no alignment at all. It just just forwarded to the C compiler.

**Comment by Araq** [comment]
> Currently, from what I can tell, the generated layout is mostly c-compatible by happy accident

Not everything in Nim that works does so "by happy accident".

**Comment by arnetheduck** [comment]
> {.packed.} simply means no alignment at all. It just just forwarded to the C compiler.

"packed" is a very specific alignment where there are no alignment padding bytes and fields are guaranteed to be in a particular order - if nim is indeed a compiler and not a transpiler, it should be able to stand on its own, in terms of its documented functionality (or explicitly defer to another documented behavior) - "whatever the c backend happens to spit out for the local c compiler" isn't a sufficiently high bar. 

the c backend has to choose c-compiler-specific directives to achieve this, and make sure that fields are in the right order - there also cannot be any type info pointers and a bunch of other things.. 

the alternative is to gut nim's ability to make better choices than C when it comes to memory layout by explicitly specifying that all nim objects that are simple enough follow the C layout / alignment.. that would be sad because one can do better.

> Not everything in Nim that works does so "by happy accident".

happy accident in my book is anything that's not in the manual or otherwise documented as intentional - if it's in your head it's a happy accident you haven't been hit by a bus.. unless "the compiler is the spec" which I certainly hope is not the case, based on the number of open issues :) the litmus test here is someone that wants to come up with an independent implementation.

**Comment by krux02** [comment]
woohoo, the tests finally run through.

**Comment by mratsim** [comment]
Congrats! I've been waiting for this for a long-time

I'd like to add the following (taken from [Stint bigints](https://github.com/status-im/nim-stint/blob/master/stint/private/datatypes.nim)) to the test suite, currently I have to implement my own `getSize` macro:

```Nim
import macros

macro uintImpl*(bits: static[int]): untyped =
  assert (bits and (bits-1)) == 0, $bits & " is not a power of 2"
  assert bits >= 8, "The number of bits in a should be greater or equal to 8"

  if bits >= 128:
    let inner = getAST(uintImpl(bits div 2))
    result = newTree(nnkBracketExpr, ident("UintImpl"), inner)
  elif bits == 64:
    result = ident("uint64")
  elif bits == 32:
    result = ident("uint32")
  elif bits == 16:
    result = ident("uint16")
  elif bits == 8:
    result = ident("uint8")
  else:
    error "Fatal: unreachable"

proc getSize*(x: NimNode): static[int] =
  var multiplier = 1
  var node = x.getTypeInst

  while node.kind == nnkBracketExpr:
    assert eqIdent(node[0], "UintImpl"), (
      "getSize only supports primitive integers, Stint and Stuint")
    multiplier *= 2
    node = node[1]

  # node[1] has the type
  # size(node[1]) * multiplier is the size in byte

  # For optimization we cast to the biggest possible uint
  result =  if eqIdent(node, "uint64") or eqIdent(node, "int64"): multiplier * 64
            elif eqIdent(node, "uint32") or eqIdent(node, "int32"): multiplier * 32
            elif eqIdent(node, "uint16") or eqIdent(node, "int16"): multiplier * 16
            elif eqIdent(node, "uint8") or eqIdent(node, "int8"): multiplier * 8
            elif eqIdent(node, "int") or eqIdent(node, "uint"):
              multiplier * 8 * sizeof(int)
            else:
              assert false, "Error when computing the size. Found: " & $node
              0

macro getSize*(x: typed): untyped =
  let size = getSize(x)
  result = quote do:
    `size`

type
  # ### Private ### #
  BaseUint* = UintImpl or SomeUnsignedInt

  UintImpl*[Baseuint] = object
    when system.cpuEndian == littleEndian:
      lo*, hi*: BaseUint
    else:
      hi*, lo*: BaseUint

  StUint*[bits: static[int]] = object
    data*: uintImpl(bits)


var a: Stuint[256]

static: echo a.data.getSize # 256
doAssert a.sizeof == 32 # With static
```

File: compiler/semmagic.nim [hunk 3]
```diff
@@ -326,13 +326,19 @@ proc magicsAfterOverloadResolution(c: PContext, n: PNode,
       # TODO there is no proper way to find out if a type cannot be queried for the size.
       let size = getSize(c.config, n[1].typ)
       # We just assume here that the type might come from the c backend
-      if size > 0:
+      if size == szImportedType:
+        # Forward to the c code generation to emit a `sizeof` in the C code.
+        result = n
+      elif size >= 0:
```
Comment (timotheecour) [inline compiler/semmagic.nim]: will that work with incomplete types that have at least 1 field? 
```
type
  MyStruct {.importc: "MyStruct".} = object
    field1: int
    # missing field2
```

see also https://github.com/nim-lang/Nim/issues/9250 for a more robust solution (that can be implemented after your PR)

Comment (krux02) [reply compiler/semmagic.nim]: originally I was looking at the wrong location for the importc flag. Now it should work properly.

Comment (timotheecour) [reply compiler/semmagic.nim]: @krux02 is my assessment in https://github.com/nim-lang/Nim/issues/9250 correct that currently you must pessimistically reject computing sizeof for any type annotated as `importc` ? feel free to comment in https://github.com/nim-lang/Nim/issues/9250

Comment (krux02) [reply compiler/semmagic.nim]: you are correct.

**Comment by timotheecour** [comment]
@krux02 
github says: "Conflicting files"
btw, instead of a giant merge that will mess up very old git history for everyone, would the following be ok:
* either attempt `git rebase -i` (potentially painful in this case, I don't know)
* or, "manual squash + rebase", by locally, merge against devel, then create a single commit based on the diff from devel

**Comment by krux02** [comment]
@Araq this is the test that is currently failing:
https://github.com/krux02/Nim/blob/sizeof-alignof/tests/typerel/trectuples.nim#L6

Can you explain me, why this should be an illegal recursion?

**Comment by timotheecour** [comment]
> Can you explain me, why this should be an illegal recursion?

odd indeed given that these are invalid:
`type Node = tuple[left: ref Node]`
and 
`type A = tuple[B: ptr A]`

given that these are valid:
```
  type A = object
    B: ptr A

  type A = object
    B: ref A
```

also, the test in https://github.com/krux02/Nim/blob/sizeof-alignof/tests/typerel/trectuples.nim#L6 is not great since the lines after line 6 are not processed

**Comment by krux02** [comment]
Recursive types need to be nominal types. Otherwise type equality checks need to have cycle detection, and that is expensive. Forbidding that kind of recursive types isn't really a big limitation. The reason for this limitation could be documented better though.

**Comment by timotheecour** [comment]
=> squashed + rebased against devel in https://github.com/nim-lang/Nim/pull/9356 to allow fast fwd merge

**Pull request closed without merging**